staging/src/k8s.io/apiserver/pkg/apis/config/validation/validation.go

```diff
@@ -78,13 +79,16 @@ func ValidateEncryptionConfiguration(c *config.EncryptionConfiguration) field.Er
 			allErrs = append(allErrs, field.Required(p, fmt.Sprintf(atLeastOneRequiredErrFmt, p)))
 		}
 
+		// kmsProviderNames is used to track config names to ensure they are unique.
+		kmsProviderNames := make(map[string]bool)
 		for j, provider := range conf.Providers {
 			path := p.Index(j)
 			allErrs = append(allErrs, validateSingleProvider(provider, path)...)
 
 			switch {
 			case provider.KMS != nil:
-				allErrs = append(allErrs, validateKMSConfiguration(provider.KMS, path.Child("kms"))...)
+				allErrs = append(allErrs, validateKMSConfiguration(provider.KMS, path.Child("kms"), kmsProviderNames)...)
```

Comment:
does this apply just to v2 or also to v1? what is the current behavior if you have multiple kmsv1 providers with the same name? is there any reason you would want to do that (e.g. during key rotation)?

Reply:
cc @enj

Reply:
I did think about making this v2 only because it's a breaking change but I think uniqueness would need to be across v1 and v2 names.

> what is the current behavior if you have multiple kmsv1 providers with the same name? is there any reason you would want to do that (e.g. during key rotation)?

All the kms providers will be configured fine without any error. Having multiple kms providers name doesn't offer any real value. For rotation, the provider that's currently first in the list will be moved down the list retaining the same name. The new kms provider that'll be used for write will be added first to the list and doesn't need the same name. Only requirement is to retain the name for the kms provider when it's moved around in the list of kms providers.

Reply:
doesn't the name factor into the prefix added to the encrypted data?

https://github.com/kubernetes/kubernetes/blob/367f01f62d1a6ca149c317f97a958dd5c6e74af1/staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/config.go#L428-L442

https://github.com/kubernetes/kubernetes/blob/367f01f62d1a6ca149c317f97a958dd5c6e74af1/staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/config.go#L488

if someone had two kms v1 providers with the same name, wouldn't requiring them to change the name of one of them in the config break reading data possibly already in etcd with that prefix?

Reply:
If we constrain this to just kms v2, there are no backward compatibility issues.  For kms v1, there is no reason to use the same name for distinct kms plugins, though technically it is "okay" (#82434 was aimed at fixing the file based providers but it enabled overlapping kms plugins as well).  Even during rotation with kms v1 I do not think there is any benefit outside of "being lazy."  But it is totally possible that someone is relying on using distinct kms plugins under the same name to decrypt data that is encrypted by both plugins.  It does mean that they are taking a performance hit whenever the DEK cache is missing the key and the data is encrypted by a plugin not at index 0.  I am unsure if you had AES-CBC based data if this would work correctly since I do not recall if that does the right integrity checking... basically I think its a bad idea to let people keep doing this.  We could do an action required release note to ask folks to run storage migration before upgrading?

Reply:
The last time we discussed this on sig-auth, I think the plan was to add an action item in the release note to run storage migration. I can update the release notes with that.

Reply:
> The last time we discussed this on sig-auth, I think the plan was to add an action item in the release note to run storage migration. I can update the release notes with that.

If you currently had two kmsv1 entries with the same name that were encrypting differently, that certainly weird and hurts performance, but to do the storage transformation, don't you have to keep both identically named kmsv1 entries as decrypting entries and add a third distinctly named entry at the beginning as an encrypting entry?

Preventing them from starting the server with the duplicate kmsv1 entries would keep them from being able to run the storage migration.

I think we should limit the unique name requirement to kmsv2 (we can detect/print nasty warnings if kmsv1 entries have duplicate names)

Reply:
I think the gist of what Jordan is saying is that you would have to perform the storage migration before upgrading.  Once you have upgraded, the API server would fail to start and fixing the config would make the encrypted data "unreachable" by the fixed plugins due to the prefix checks including the plugin name.  Requiring a downgrade to fix the issue is probably beyond what we would consider a reasonable ask.  Let's just fix v2 plugins.

Reply:
I think we want to make sure that v2 plugins have a name that does not conflict with any other v1 or v2 plugins.

After the reloading flag PR is merged, we should require all plugins to have a unique name when reloading is enabled because that will make it easier to re-use the DEK cache across reloads.  It's a bit weird to change validation behavior based on a flag, but it's backwards compatible and you aren't forced to use reloading...

Reply:
done!

Reply:
done!